# HG changeset patch
# User William A. Kennington III <william@wkennington.com>
# Date 1449178027 28800
#      Thu Dec 03 13:27:07 2015 -0800
# Node ID 190fb1c96de8dd32e7239e62b8a94ab06a7b17b3
# Parent  1a2e8b6e713c728220ad340c871e49e8290a0e0b
Allow nix to properly load shared libraries.

diff -r 1a2e8b6e713c -r 190fb1c96de8 cmd/shlibsign/shlibsign.c
--- a/cmd/shlibsign/shlibsign.c	Thu Dec 03 17:38:49 2015 +1100
+++ b/cmd/shlibsign/shlibsign.c	Thu Dec 03 13:27:07 2015 -0800
@@ -862,6 +862,8 @@
     libname = PR_GetLibraryName(NULL, "softokn3");
     assert(libname != NULL);
     lib = PR_LoadLibrary(libname);
+    if (!lib)
+        lib = PR_LoadLibrary(NIX_NSS_LIBDIR"libsoftokn3.so");
     assert(lib != NULL);
     PR_FreeLibraryName(libname);
 
diff -r 1a2e8b6e713c -r 190fb1c96de8 coreconf/config.mk
--- a/coreconf/config.mk	Thu Dec 03 17:38:49 2015 +1100
+++ b/coreconf/config.mk	Thu Dec 03 13:27:07 2015 -0800
@@ -188,3 +188,6 @@
 
 # Hide old, deprecated, TLS cipher suite names when building NSS
 DEFINES += -DSSL_DISABLE_DEPRECATED_CIPHER_SUITE_NAMES
+
+# Nix specific stuff.
+DEFINES += -DNIX_NSS_LIBDIR=\"$(out)/lib/\"
diff -r 1a2e8b6e713c -r 190fb1c96de8 lib/pk11wrap/pk11load.c
--- a/lib/pk11wrap/pk11load.c	Thu Dec 03 17:38:49 2015 +1100
+++ b/lib/pk11wrap/pk11load.c	Thu Dec 03 13:27:07 2015 -0800
@@ -429,6 +429,13 @@
 	 * unload the library if anything goes wrong from here on out...
 	 */
 	library = PR_LoadLibrary(mod->dllName);
+	if ((library == NULL) &&
+	    !rindex(mod->dllName, PR_GetDirectorySeparator())) {
+            library = PORT_LoadLibraryFromOrigin(my_shlib_name,
+                                      (PRFuncPtr) &softoken_LoadDSO,
+                                      mod->dllName);
+	}
+
 	mod->library = (void *)library;
 
 	if (library == NULL) {
diff -r 1a2e8b6e713c -r 190fb1c96de8 lib/util/secload.c
--- a/lib/util/secload.c	Thu Dec 03 17:38:49 2015 +1100
+++ b/lib/util/secload.c	Thu Dec 03 13:27:07 2015 -0800
@@ -69,9 +69,14 @@
 
     /* Remove the trailing filename from referencePath and add the new one */
     c = strrchr(referencePath, PR_GetDirectorySeparator());
+    if (!c) { /* referencePath doesn't contain a / means that dladdr gave us argv[0]
+               * and program was called from $PATH. Hack to get libs from NIX_NSS_LIBDIR */
+        referencePath = NIX_NSS_LIBDIR;
+        c = (char*) &referencePath[sizeof(NIX_NSS_LIBDIR) - 1]; /* last / */
+    }
     if (c) {
         size_t referencePathSize = 1 + c - referencePath;
-        fullName = (char*) PORT_Alloc(strlen(name) + referencePathSize + 1);
+        fullName = (char*) PORT_Alloc(strlen(name) + referencePathSize + 5);
         if (fullName) {
             memcpy(fullName, referencePath, referencePathSize);
             strcpy(fullName + referencePathSize, name); 
@@ -81,6 +86,11 @@
 #endif
             libSpec.type = PR_LibSpec_Pathname;
             libSpec.value.pathname = fullName;
+            if ((referencePathSize >= 4) &&
+                (strncmp(fullName + referencePathSize - 4, "bin", 3) == 0)) {
+                memcpy(fullName + referencePathSize -4, "lib", 3);
+            }
+            strcpy(fullName + referencePathSize, name);
             dlh = PR_LoadLibraryWithFlags(libSpec, PR_LD_NOW | PR_LD_LOCAL
 #ifdef PR_LD_ALT_SEARCH_PATH
             /* allow library's dependencies to be found in the same directory
@@ -88,6 +98,10 @@
                                           | PR_LD_ALT_SEARCH_PATH 
 #endif
                                           );
+            if (! dlh) {
+                strcpy(fullName + referencePathSize, name);
+                dlh = PR_LoadLibraryWithFlags(libSpec, PR_LD_NOW | PR_LD_LOCAL);
+            }
             PORT_Free(fullName);
         }
     }
